deps/media-playback/media-playback/media.c

```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
+		break;
+	}
+}
 
-			if (!mp_media_prepare_frames(m))
-				return false;
-			if (mp_media_eof(m))
-				continue;
+/*Executes the top message in the queue.*/
+static bool mp_media_pop_message(mp_media_t *m)
+{
+	bool result = true;
+
+	mp_media_queue_node_t *front = m->mp_queue.first;
+	/*Only execute if a state change actually occurs. If an
+	  action needs to be called within the state, add a new
+	  state that executes the action then moves back.*/
+	if (m->mp_state != front->message.state) {
+		result = front->message.state_handler(m);
+	}
+
+	pthread_mutex_lock(&m->mutex);
+
+	if (m->mp_queue.first->next != NULL)
+		m->mp_queue.first = front->next;
+	else {
+		m->mp_queue.first = NULL;
+		m->mp_queue.last = NULL;
+	}
+	m->mp_queue.flags -= front->message.state;
+	free(front);
+
+	pthread_mutex_unlock(&m->mutex);
+
+	return result;
+}
+
+static bool mp_media_push_message(mp_media_t *m, enum mp_media_state state)
+{
+	/*To prevent the queue from growing infinitely, and to
+	  prevent needless repetition, use a bit flag.
+	  Queue should only contain any given message once.*/
+	if (m->mp_queue.flags & state)
+		return false;
+
+	mp_media_queue_node_t *newnode = malloc(sizeof(mp_media_queue_node_t));
+	mp_media_state_message_t *newmessage;
+
+	newmessage = &(newnode->message);
+
+	newmessage->state = state;
+
+	/*Register event handlers here*/
+	switch (state) {
+	case mstate_playing:
+		newmessage->state_handler = &mp_media_playing;
+		break;
+	case mstate_resetting:
+		newmessage->state_handler = &mp_media_resetting;
+		break;
+	case mstate_setloop:
+		newmessage->state_handler = &mp_media_setloop;
+		break;
+	case mstate_setreconnect:
+		newmessage->state_handler = &mp_media_setreconnect;
+		break;
+	case mstate_paused:
+		newmessage->state_handler = &mp_media_pausing;
+		break;
+	case mstate_resume:
+		newmessage->state_handler = &mp_media_resume;
+		break;
+	case mstate_stopping:
+		newmessage->state_handler = &mp_media_stopping;
+		break;
+	case mstate_stopped:
+		newmessage->state_handler = &mp_media_stopped;
+		break;
+	case mstate_seeking:
+		newmessage->state_handler = &mp_media_seeking;
+		break;
+	case mstate_killing:
+		newmessage->state_handler = &mp_media_killing;
+		break;
+	default:
+		break;
+	}
+	newnode->next = NULL;
+
+	pthread_mutex_lock(&m->mutex);
+
+	if (m->mp_queue.last == NULL)
+		m->mp_queue.first = newnode;
+	else
+		m->mp_queue.last->next = newnode;
+
+	m->mp_queue.last = newnode;
+	m->mp_queue.flags |= state;
+
+	pthread_mutex_unlock(&m->mutex);
+
+	return true;
+}
+
+/*Manage loop timing.*/
+static inline bool mp_media_tick(mp_media_t *m)
+{
+	/*If there is an event that should block the loop, place it
+	  here. Make sure to address semaphore logic in all events.
+	  For instance, a stop invoke should check for a pause event
+	  and signal only then. If stop always signals, then that
+	  signal may prevent the block for stop itself, but without
+	  signalling out of a pause, the stop event will not run until
```

Comment:
Not a big deal, but "signaling" with one "l" is the more commonly accepted American English spelling.